Update more samplers for v4 compatibility

brandonwillard opened the pull request:
This PR makes the changes necessary to run more step methods in v4 (e.g. basic Metropolis, slice sampler, etc.)

brandonwillard committed:
Make Metropolis, Slice, PGBART, MetropolisMLDA use point values

brandonwillard committed:
Re-enable disabled tests

michaelosthege commented on .github/workflows/pytest.yml:
```diff
@@ -27,27 +27,27 @@ jobs:
           - |
             --ignore=pymc3/tests/test_distribution_defaults.py
-            --ignore=pymc3/tests/test_distributions.py
+            # --ignore=pymc3/tests/test_distributions.py
```
Unfortunately we can't use comments here. That's why all the comments had to be placed above.

michaelosthege commented on pymc3/step_methods/metropolis.py:
```diff
@@ -212,15 +211,18 @@ def astep(self, q0):
                 delta = np.round(delta, 0).astype("int64")
                 q0 = q0.astype("int64")
-                q = (q0 + delta).astype("int64")
+                q = (q0.data + delta).astype("int64")
```
Did the type of `q0` change? If so it should probably be added as a type hint in the signature.

michaelosthege commented on pymc3/step_methods/slicer.py:
```diff
@@ -61,36 +62,40 @@ def __init__(self, vars=None, w=1.0, tune=True, model=None, iter_limit=np.inf, *
-            while y <= logp(ql):  # changed lt to leq  for locally uniform posteriors
+            while y <= logp(
+                RaveledVars(ql, q0.point_map_info)
+            ):  # changed lt to leq  for locally uniform posteriors
```
The line is split by `black` only because of the comment (compare next `while`). The comment should probably go on its own line above..

michaelosthege commented on pymc3/step_methods/slicer.py:
```diff
@@ -61,36 +62,40 @@ def __init__(self, vars=None, w=1.0, tune=True, model=None, iter_limit=np.inf, *
             cnt = 0
             q[i] = nr.uniform(ql[i], qr[i])
-            while logp(q) < y:  # Changed leq to lt, to accomodate for locally flat posteriors
+            while logp(q_ra) < y:  # Changed leq to lt, to accomodate for locally flat posteriors
```
⚠ before this line used `q`, which was inplace-edited in the line above.
If and only if `RaveledVars` maintains the pointer to the underlying `ndarray`, the `RaveledVars` objects should be created all right after the `copy()` lines (L69).

brandonwillard replied:
A `RaveledVars` object is just a named tuple, so, yes, it holds a reference to `q` and reflects any in-place changes made to `q`.

michaelosthege reviewed (commented):
Let me know if I can help to take care of the remaining metropolis samplers - particularly w.r.t. `RaveledVars` changes.

brandonwillard committed:
Use value vars to determine steps

brandonwillard committed:
Use size instead of shape in pymc3.tests.sampler_fixtures

brandonwillard committed:
Set model-level RandomVariable seeds during sampling

brandonwillard committed:
Check shapes by evaluating graph with start values

brandonwillard committed:
Fix logpt so that transforms are always applied, when enabled

brandonwillard committed:
Set default transform for Dirichlet

brandonwillard committed:
Normalize Multinomial argument

brandonwillard committed:
Use no_transform_object in Distribution.__new__

brandonwillard committed:
Fix Interval.jacobian_det

brandonwillard committed:
Fix Stickbreaking scalar condition

brandonwillard committed:
Make Model.test_point generation transform existing RV test value

brandonwillard committed:
Add transformed value variables to Model.named_vars

brandonwillard committed:
Remove DeterministicWrapper from Deterministic

brandonwillard committed:
Make sure sample_posterior_predictive doesn't use trace values for sampled vars

brandonwillard committed:
Set seed after loading trace in TestSaveLoad

brandonwillard committed:
Make v4 compatibility changes to pymc3.tests.test_sampling

brandonwillard committed:
Make pymc3.tests.test_transforms work with None RV variables

brandonwillard committed:
Make find_MAP work with RaveledVars

brandonwillard committed:
Make sure start values are NumPy arrays

brandonwillard committed:
Raise NotImplementedError in Group.__init__

brandonwillard committed:
Add type hints to astep methods

brandonwillard committed:
Use untransformed samples and xfail Arviz tests in BaseSampler

brandonwillard committed:
Make sure forward transformed input is a TensorVariable in TestMatchesScipy

brandonwillard committed:
Adjust flaky last-digit numerical requirements in TestMatchesScipy

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1075,6 +1075,7 @@ def test_beta(self):
             {"alpha": Rplus, "beta": Rplus},
             lambda value, alpha, beta: sp.beta.logcdf(value, alpha, beta),
             n_samples=10,
+            decimal=5,
```
This will probably lead to issues in float32 runs. Usually you have to do `decimal=select_by_precision(float64=x, float32=y)`. I had to do something similar for the logp test in my PR https://github.com/pymc-devs/pymc3/pull/4548/commits/59950dd75bea0a185634ced24768964168001a26

brandonwillard replied:
Ah, all right, I'll change it to something like that.

brandonwillard commented:
@ricardoV94, once these tests pass, I'll merge this and help out with #4548 (e.g. it will need to be rebased, etc.)

brandonwillard committed:
Fix MvNormal quaddist_matrix parameter order

brandonwillard committed:
Enable MvNormal tests in test_distributions

brandonwillard committed:
Factor out parameter pre-processing in TestMatchesScipy

brandonwillard committed:
Apply recent xfail updates from master branch

brandonwillard committed:
Prevent dtype conversion in Aesara during testing to avoid a bug

This change avoids the bug fixed in
https://github.com/pymc-devs/aesara/pull/348.

brandonwillard committed:
Fix NegativeBinomial parameterization and enable its tests

brandonwillard committed:
Prevent SciPy error by using float64 point in test_dirichlet_with_batch_shapes
